flax/nnx/nn/recurrent.py

```diff
@@ -0,0 +1,920 @@
+        self.dense_i = Linear(
+            in_features=in_features,
+            out_features=3 * hidden_features,  # r, z, n
+            use_bias=True,
+            kernel_init=self.kernel_init,
+            bias_init=self.bias_init,
+            dtype=self.dtype,
+            param_dtype=self.param_dtype,
+            rngs=rngs,
+        )
+
+        self.dense_h = Linear(
+            in_features=hidden_features,
+            out_features=3 * hidden_features,  # r, z, n
+            use_bias=False,
+            kernel_init=self.recurrent_kernel_init,
+            dtype=self.dtype,
+            param_dtype=self.param_dtype,
+            rngs=rngs,
+        )
+
+    def __call__(self, carry, inputs):
+        """Gated recurrent unit (GRU) cell.
+
+        Args:
+            carry: the hidden state of the GRU cell,
+              initialized using ``GRUCell.initialize_carry``.
+            inputs: an ndarray with the input for the current time step.
+              All dimensions except the final are considered batch dimensions.
+
+        Returns:
+            A tuple with the new carry and the output.
+        """
+        h = carry
+
+        # Compute combined transformations for inputs and hidden state
+        x_transformed = self.dense_i(inputs)
+        h_transformed = self.dense_h(h)
+
+        # Split the combined transformations into individual components
+        xi_r, xi_z, xi_n = jnp.split(x_transformed, 3, axis=-1)
+        hh_r, hh_z, hh_n = jnp.split(h_transformed, 3, axis=-1)
+
+        # Compute gates
+        r = self.gate_fn(xi_r + hh_r)
+        z = self.gate_fn(xi_z + hh_z)
+
+        # Compute n with an additional linear transformation on h
+        n = self.activation_fn(xi_n + r * hh_n)
+
+        # Update hidden state
+        new_h = (1.0 - z) * n + z * h
+        return new_h, new_h
+
+    def initialize_carry(
+        self, rngs: rnglib.Rngs | None, input_shape: tuple[int, ...]
+    ) -> Array:
+        """Initialize the RNN cell carry.
+
+        Args:
+            rngs: random number generator passed to the init_fn.
+            input_shape: a tuple providing the shape of the input to the cell.
+
+        Returns:
+            An initialized carry for the given RNN cell.
+        """
+        batch_dims = input_shape[:-1]
+        if rngs is None:
+            rngs = self.rngs
+        mem_shape = batch_dims + (self.hidden_features,)
+        h = self.carry_init(rngs.params(), mem_shape, self.param_dtype)
+        return h
+
+    @property
+    def num_feature_axes(self) -> int:
+        return 1
+
+
+class RNN(Module):
+    """The ``RNN`` module takes any :class:`RNNCellBase` instance and applies it over a sequence
+
+    using :func:`flax.linen.scan`.
+    """
+
+    def __init__(
+        self,
+        cell: RNNCellBase,
+        time_major: bool = False,
+        return_carry: bool = False,
+        reverse: bool = False,
+        keep_order: bool = False,
+        unroll: int = 1,
```

Comment:
Maybe we could accepts `rngs` to get a default that can be optionally override during `__call__`?
```suggestion
        unroll: int = 1,
        rngs: rnglib.Rngs | None = None,
```

Reply:
changed and set default to rngs(0)